### ISSUE TITLE
Possible addition of centroidization in XCMS

### ISSUE THREAD

**Issue by adelabriere** [sneumann/xcms#268]
Hello, 

I developped a centroidization algorithm adapted to high resolution MS/MSn which takes as input and xcmRaw object and return a centroided xcmsRaw object. I feel like this could be a good addition to xcms.

It was tested on 3 different mass spectrometers TOF, Exactive, Fusion in LC and FIA-MS and worked fine on every use case, with better results than Proteowizard centroidization.

However the algorithm is written in R/C++, and it uses somes Rcpp/C++ features. At the moment it is a simple script which use Rcpp and Inline, and  need to be sourced before every use, therefore there is some work for the transcription in XCMS, especially to write the C++ in C, 

It may be found here : https://github.com/adelabriere/centroidizationXraw/blob/master/centroidizationScript.R

It is quite raw, but it have worked well during my PhD, and it does not require any parameter in the general case, if you are interested I can try to convert it to a cleaner/more adapted form for xcms.

In this case some advices will be appreciated about the step to do so will be appreciated, I imagine that I need to define a new S4 method.generic for xcmsRaw and put it there but I have no idea about the rest (Naming/Coding convention etc...)

Thanks,
Alexis

**Comment by adelabriere** [comment]
I'll try to stick to the new interface then, thanks !

**Comment by jotsetung** [comment]
Is the way you do the centroiding specific to metabolomics data? if it was applicable to MS data in general I think it might better go to the `MSnbase` package instead of `xcms`. The new `xcms` interface @sneumann mentioned reuses classes from `MSnbase` directly, specifically the `OnDiskMSnExp` (which replaces the `xcmsRaw`).

Note that we recently (means last week :) ) added some centroiding functionality to the `MSnbase` package (it's in the current developmental version at https://github.com/lgatto/MSnbase - see https://bioconductor.org/packages/devel/bioc/vignettes/MSnbase/inst/doc/MSnbase-centroiding.html for the vignette describing the centroiding).

A little guidance as starters: centroiding could/should be done on a per-Spectrum level. the function/method that does the centroiding is called `pickPeaks` (https://github.com/lgatto/MSnbase/blob/master/R/functions-Spectrum.R#L404:L458). this is called on the profile data of each spectrum and is supposed to return a centroided spectrum.

I'm really curious what your centroiding does. Could you share some info on your approach?

Let me also know if you need more help.

**Comment by adelabriere** [comment]
I think it's quite similar to your, from what I see. It include the follwoing 2 steps :

*Determination of mz step function
*Peak detection using local maxima/minima

I start by determing the m/z sampling interval in function of m/z, to do so i take the scan at the most intense position on the TIC and I caculate the m/z difference ebtween every consecutive points (x[i]- x[i-1]). This is then smoothed by median filtering. This give a straight line if the file is centroided, on which a linear regression is performed. If the regression fit badly it's that the file is already centroided. Without that I have a good estimate of the step function.

If it fits well ,each scan is processed independently using this step function. Each (mz,int) point is processed sequentially in increasing m/z order. There is always a buffer containing the potential peak :

If the intensity is coherent with the position in the peak (before after the maximum), the point is added to the peak, if it not, or if the mz different with the previous point exceed 2 time (frac parameter) the mass step function, it is put in a new potential peak, and the old peak is inserted in the centroided peak list.

The m/z is always refined by weighted mean (that was the main issue with proteowizard peak picking)

Therefore I don't know if my approach is complementary/worst/better. I think that I need to compare the two approaches before making any code refactoring. 

Eventually a better centroidization may be attained by filtering in mass dimension as you did and incorporating the mass step function as I did(not sure about that though).

I'll try to compare the two methods. On which file was your vignette generated ?

**Comment by jotsetung** [comment]
Thanks for the info

> The m/z is always refined by weighted mean (that was the main issue with proteowizard peak picking)

Yes, that's what the `refineMz` functions also do - I guess the main difference is the way how the peak area is defined, i.e. which data points are considered for the m/z refinement. Here your method sounds to do it differently.

> I'll try to compare the two methods. On which file was your vignette generated ?

That would be awesome! The vignette was performed on a file within the `sciex` folder within the `msdata` package - You'll need R-3.5-devel and install on that the developmental Bioconductor packages:
```r
source("https://bioconductor.org/biocLite.R")
library(BiocInstaller)
biocLite(c("MSnbase", "msdata"))
```

**Comment by stanstrup** [comment]
Would be nice to also compare to the vendor centroiding.

**Comment by sneumann** [comment]
And to OpenMS. Question: what is *good* centroiding ? How to decide/calculate the quality ?
I always love to conceive a way to create a data-driven untargeted evaluation. 
Are there QC values that are influenced by the quality of the centroidisation ?
https://www.ncbi.nlm.nih.gov/pubmed/24760958

**Comment by stanstrup** [comment]
some fast ideas: 

1) accuracy of know compounds
2) detection of isotope peaks as a way to assess ability to detect small peaks. so for example count all peaks below some intensity that fit with an isotope.

**Comment by jotsetung** [comment]
> Would be nice to also compare to the vendor centroiding.

I did that for AB Sciex - they're essentially just reporting the m/z and intensity of the maximum signal. ~ same results as doing a simple `pickPeaks` with the raw, un-smoothed data. Have no other data on hand, so I couldn't check for other manufacturers.

Regarding *decide on quality of centroiding*: to compare approaches I did:
+ (as @stanstrup suggests above) evaluate accuracy of the centroided data for a set of ~ 60 *known compounds* (difference between theoretical m/z and m/z reported for the identified chromatographic peak).
+ m/z width of the detected chromatographic peaks. I used this as a proxy for the reduction in m/z scattering between consecutive scans, i.e. if the MS detects an ion, the scattering of values around the true m/z should be ideally as low as possible.

**Comment by sneumann** [comment]
I have created https://goo.gl/dQdpwh. I'd like to ask people to plonk the centroided versions
using different centroidisers there ? I'll give write access if you send your google account my way.

**Comment by SiggiSmara** [comment]
And as far as I can tell proteowizard (take a look at the CwtPeakDetector::refinePeaks function in the msconvert code) is only using continuous waveform to detect the rough location of a spectral peak and then using a simple local maximum to assign the centroid location.

**Comment by SiggiSmara** [comment]
Agree with @stanstrup that vendor centroiding should be compared as well but for vendor centroiding you also have to be careful. In at least some of the more recent proteowizard there is no access to waters centroiding and in at least one instance the gui version of the msconvert was not even complaining about it even if you selected vendor centroiding.

**Comment by stanstrup** [comment]
@SiggiSmara that is right. But in masslynx you can manually centroid a raw file and then convert the resulting centroided raw file.

**Comment by SiggiSmara** [comment]
@stanstrup that is for sure a much better approach. I have to look and see if Sciex Peakview can do that. So far I've only converted individual spectra. In any case as far as I can tell Peakview is doing local maxima centroiding, you can apply smoothing functions before but that you have to do manually as well.

**Comment by adelabriere** [comment]
@SiggiSmara  No you don't miss anything, the data in orbitrap and Exactive are really clean, but they are not really centroided ( ![ex_peak_exactive](https://user-images.githubusercontent.com/17563377/37708519-426634d2-2d07-11e8-90d7-ae8a542043ec.png)).
A centroidization implying mz refinement as the one developped by @jotsetung  and my centroidization still improve the accuracy of the m/z compared to picking the most intense peak m/z. Moreover in some case the point with the highest intensity change between consecutives scans, resulting in 2 mass traces or 0 during ROI/mass trace detection if the parametrization is not very accurate. 

Therefore for the majority of signal raw peak picking is sufficient,but you'll miss some information compared to a routine implying mz refinment.

**Comment by SiggiSmara** [comment]
@adelabriere, no I can see that my assumption was wrong looking at the data. I was assuming that the FT resulted in much more centroid like data.

**Comment by stanstrup** [comment]
@adelabriere That image is heavily cropped on the intensity scale, right? The highest peak is many times larger than what we see?
 
The orbitrap data I have seen is rarely symmetric like that and thus centroiding that further I think would not give correct results.
MzMine fits a Gaussian of some kinds and removes peaks with intensities below this Gaussian (from what I remember from a conversation with the author).
I have made a filter that starts from the highest peak and removes small peaks around it (https://github.com/stanstrup/chemhelper/blob/master/R/orbi.filter.R) to do the same thing.
